datadog/dogstatsd/base.py

```diff
@@ -420,14 +437,21 @@ def _is_telemetry_flush_time(self):
     def _send_to_server(self, packet):
         flush_telemetry = self._is_telemetry_flush_time()
         if flush_telemetry:
-            packet += "\n"+self._flush_telemetry()
-
+            telemetry = self._flush_telemetry()
+            same_packet = (len(telemetry) + len(packet) < self._max_payload_size)
+            if same_packet:
+                packet += "\n"+telemetry
```

Comment:
Why not compute the length of the telemetry at startup and decrease `_max_payload_size` by it (we need the tags first and can set a high estimate for the metrics values). This would simplify the logic but more importantly avoid sending multiple packet when flushing the telemetry. It's what we do in the ruby client.

When looking at the size of the telemetry payload compare to a regular metric is will never fit in the remaining space of a buffer. This means that we will always have two `socket.send` in the next `try/except` that can be raise an exception. Also, currently the metrics will be wrong if we succeed in sending the metrics but fail for the telemetry packet (we would not increase `packets_sent` and `bytes_sent` line 453.

Reply:
If we systematically lower the max buffer size by the telemetry size, we may increase the overall packet rate because packet without telemetry won't be filled (almost half the payload for UDP). I guess it depends on the submission pattern and as discussed before I don't have any number regarding packet size. I think anyhow that reworking the byte accounting will make the code much readable.